File: app/src/main/java/de/grobox/transportr/trips/detail/StopViewHolder.java
package de.grobox.transportr.trips.detail;


import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import de.grobox.transportr.R;
import de.schildbach.pte.dto.Stop;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static de.grobox.transportr.utils.TransportrUtils.getLocationName;

class StopViewHolder extends BaseViewHolder {

	private final ImageView circle;
	private final TextView stopLocation;
	private final ImageButton stopButton;

	StopViewHolder(View v) {
		super(v);
		circle = v.findViewById(R.id.circle);
		stopLocation = v.findViewById(R.id.stopLocation);
		stopButton = v.findViewById(R.id.stopButton);
	}

	void bind(Stop stop, LegClickListener listener, int color) {
		if (stop.getArrivalTime() != null) {
			setArrivalTimes(fromTime, fromDelay, stop);
		} else {
			fromDelay.setVisibility(GONE);
		}

		if (stop.getDepartureTime() != null) {
			if (stop.getDepartureTime().equals(stop.getArrivalTime())) {
				toTime.setVisibility(GONE);
				toDelay.setVisibility(GONE);
			} else {
				setDepartureTimes(toTime, toDelay, stop);
				toTime.setVisibility(VISIBLE);
			}
		} else {
			toTime.setVisibility(GONE);
			toDelay.setVisibility(GONE);
		}

		circle.setColorFilter(color);

		stopLocation.setText(getLocationName(stop.location));
		stopLocation.setOnClickListener(view -> listener.onLocationClick(stop.location));

		// platforms
//		if (stop.plannedArrivalPosition != null) {
//			arrivalPlatform.setText(stop.plannedArrivalPosition.name);
//		}
//		if (stop.plannedDeparturePosition != null) {
//			departurePlatform.setText(stop.plannedDeparturePosition.name);
//		}

		// show popup on button click
		stopButton.setOnClickListener(view -> new LegPopupMenu(stopButton.getContext(), stopButton, stop).show());
	}

}

File: app/src/main/java/de/grobox/transportr/trips/detail/StopViewHolder.kt
/*
 *    Transportr
 *
 *    Copyright (c) 2013 - 2021 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.grobox.transportr.trips.detail


import android.view.View
import android.view.View.*
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import de.grobox.transportr.trips.BaseViewHolder
import de.grobox.transportr.utils.DateUtils.formatTime
import de.grobox.transportr.utils.TransportrUtils.getLocationName
import de.schildbach.pte.dto.Stop
import kotlinx.android.synthetic.main.list_item_stop.view.*
import java.util.Date


internal class StopViewHolder(v: View, private val listener : LegClickListener) : BaseViewHolder(v) {

    private val circle: ImageView = v.circle
    private val stopLocation: TextView = v.stopLocation
    private val stopButton: ImageButton = v.stopButton

    fun bind(stop: Stop, color: Int) {
        if (stop.arrivalTime != null) {
            setArrivalTimes(fromTime, fromDelay, stop)
            fromTime.visibility = VISIBLE
        } else {
            fromDelay.visibility = GONE
            if (stop.departureTime == null) {
                // insert dummy time field for stops without times set, so that stop circles align
                fromTime.text = formatTime(context, Date())
                fromTime.visibility = INVISIBLE
            } else {
                fromTime.visibility = GONE
            }
        }

        if (stop.departureTime != null) {
            if (stop.departureTime == stop.arrivalTime) {
                toTime.visibility = GONE
                toDelay.visibility = GONE
            } else {
                setDepartureTimes(toTime, toDelay, stop)
                toTime.visibility = VISIBLE
            }
        } else {
            toTime.visibility = GONE
            toDelay.visibility = GONE
        }

        circle.setColorFilter(color)

        stopLocation.text = getLocationName(stop.location)
        stopLocation.setOnClickListener { listener.onLocationClick(stop.location) }
        stopLocation.addPlatform(stop.arrivalPosition)

        // show popup on button click
        stopButton.setOnClickListener { LegPopupMenu(stopButton.context, stopButton, stop).show() }
    }

}
